### ISSUE TITLE
DataDisk does not support encryption

### ISSUE THREAD

**Issue by AdallomRoy** [Azure/azure-sdk-for-go#566]
Looks like only the OSDisk type has the encryption options type.
As far as I understand it's currently not supported to create an encrypted data disk?

**Comment by salameer** [comment]
Hello @AdallomRoy 

Thanks for reaching out...

So this is a bit tricky there  are a couple  of possibilities here:
- We're missing support for some  feature of operation on GO
- THe swagger representing the Service might be missing specs for the Encryption
- This might not be exposed for the service.

First step we'll try to understand under what bucket does this belong. if you have any more insight here please feel free to throw them

Captain @marstr will be taking a look at this

**Comment by AdallomRoy** [comment]
Hi @salameer 
I'm not sure what is supported by the REST APIs, but it is definitely supported by PowerShell, see:
https://docs.microsoft.com/en-us/azure/security/azure-security-disk-encryption

**Comment by marstr** [comment]
I did some digging, thanks for pointing out this gap in capability, @AdallomRoy. It looks like there are a handful of things that are going to need to be fixed. As can be seen [here](https://docs.microsoft.com/en-us/rest/api/compute/virtualmachines/virtualmachines-create-or-update), the capability to encrypt data disks via REST is not yet documented. Before we can get the fix into the _Azure SDK for Go_ we'll need to update the Swaggers. I'll open a new issue against Azure/azure-rest-api-specs to track this issue, and the ability to encrypt data disks will be released in Go in the next release or so.

**Comment by marstr** [comment]
I'll close out this copy of the issue, as all actionable items will be represented in the issue noted above.

**Comment by marstr** [comment]
I'm re-opening this issue until I can verify that Anu's solution from Azure/azure-rest-api-specs#1035 works with the Go SDK:
> Encryption of data disks and OS disk are done by installing 'encryption extension' in the VM, so if current swagger supports specifying extension for VM then encryption can be done.

Now, there is a more convenient way of doing this for the OS Disk, as can be seen here:
https://github.com/Azure/azure-rest-api-specs/blob/be22ed1cacc8cfa2d75e4ab8ae867ef13fb68da7/arm-compute/2016-04-30-preview/swagger/compute.json#L3656

**Comment by marstr** [comment]
I'm putting the finishing touches on an example of how to encrypt Data and/or OS disks using the Azure SDK for Go. I'll leave this issue open until that sample is published.

This process has proved to not be very straight-forward, and as of the time of writing this comment, the documentation is _sparse_ to say the least. The best article I found on it is [here](https://docs.microsoft.com/en-us/azure/virtual-machines/linux/encrypt-disks#encrypt-virtual-machine). However, now that I have it working, the code itself is pretty straight forward.

> there is a more convenient way of doing this for the OS Disk, as can be seen here: [Azure/azure-rest-api-specs@be22ed/arm-compute/2016-04-30-preview/swagger/compute.json#L3656](https://github.com/Azure/azure-rest-api-specs/blob/be22ed1cacc8cfa2d75e4ab8ae867ef13fb68da7/arm-compute/2016-04-30-preview/swagger/compute.json#L3656)

The comment that I made above is misinformed. It turns out that EncryptionSettings as a parameter in this context isn't a mechanism for creating an encrypted disk. Rather, if an OS disk is already encrypted this is the mechanism to allow Azure to boot the VM. It is assumed that once the VM is booted, the AzureDiskEncryptionExtension will be able to deal with encrypting/unencrypting all of the data disks.

That means that there is no more convenient way to encrypt disks than @anuchandy's solution. In fact, the CLI etc is installing this extension in the background.
